{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import math\n",
    "from typing import Tuple\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from scipy.stats import binom\n",
    "\n",
    "pd.set_option(\"display.max_columns\", None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "snv_file_path = \"/home/junkhann/daten/snvs_with_segments/snvs_with_segments_4100314.vcf.gz\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "81"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with open(\"header_names.txt\") as file:\n",
    "    header_names = file.readlines()\n",
    "\n",
    "header_names = [h.strip(\"\\n\") for h in header_names]\n",
    "len(header_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['#CHROM',\n",
       " 'POS',\n",
       " 'ID',\n",
       " 'REF',\n",
       " 'ALT',\n",
       " 'QUAL',\n",
       " 'FILTER',\n",
       " 'INFO',\n",
       " 'FORMAT',\n",
       " 'sample_control',\n",
       " 'sample_tumor',\n",
       " 'DBSNP',\n",
       " '1K_GENOMES',\n",
       " 'PHASE3_1K_GENOMES',\n",
       " 'ExAC_nonTCGA',\n",
       " 'LOCAL_CONTROL_2',\n",
       " 'LOCAL_CONTROL_3',\n",
       " 'LOCAL_CONTROL_4',\n",
       " 'ANNOVAR_FUNCTION',\n",
       " 'GENE',\n",
       " 'EXONIC_CLASSIFICATION',\n",
       " 'ANNOVAR_TRANSCRIPTS',\n",
       " 'SEGDUP',\n",
       " 'CYTOBAND',\n",
       " 'MAPABILITY',\n",
       " 'HISEQDEPTH',\n",
       " 'SIMPLE_TANDEMREPEATS',\n",
       " 'REPEAT_MASKER',\n",
       " 'DUKE_EXCLUDED',\n",
       " 'DAC_BLACKLIST',\n",
       " 'SELFCHAIN',\n",
       " 'CpGislands',\n",
       " 'CgiMountains',\n",
       " 'Enhancers',\n",
       " 'miRNAs_snoRNAs',\n",
       " 'miRBase18',\n",
       " 'miRNAtargets',\n",
       " 'phastConsElem20bp',\n",
       " 'TFBScons',\n",
       " 'DNMT3_BS',\n",
       " 'COSMIC',\n",
       " 'COSMIC_indels',\n",
       " 'chromosome',\n",
       " 'start',\n",
       " 'end',\n",
       " 'crest',\n",
       " 'length',\n",
       " 'tcnNbrOfLoci',\n",
       " 'tcnMean',\n",
       " 'tcnNbrOfSNPs',\n",
       " 'tcnNbrOfHets',\n",
       " 'dhNbrOfLoci',\n",
       " 'map',\n",
       " 'dhMax',\n",
       " 'cluster',\n",
       " 'neighbour',\n",
       " 'distDH',\n",
       " 'errorSNP',\n",
       " 'distTcn',\n",
       " 'errorLength',\n",
       " 'totalError',\n",
       " 'minStart',\n",
       " 'maxStart',\n",
       " 'minStop',\n",
       " 'maxStop',\n",
       " 'area',\n",
       " 'peaks',\n",
       " 'meanCovT',\n",
       " 'meanCovB',\n",
       " 'tcnMeanRaw',\n",
       " 'AF',\n",
       " 'BAF',\n",
       " 'dhMean',\n",
       " 'c1Mean',\n",
       " 'c2Mean',\n",
       " 'A',\n",
       " 'B',\n",
       " 'genotype',\n",
       " 'TCN',\n",
       " 'GNL',\n",
       " 'PID']"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with open(\"snv_usecols.txt\") as file:\n",
    "    usecols_list = file.readlines()\n",
    "\n",
    "usecols_list = [h.strip(\"\\n\") for h in header_names]\n",
    "\n",
    "usecols_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(3829904, 81)"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_snv = pd.read_csv(snv_file_path, sep='\\t', names=header_names, usecols=usecols_list, low_memory=False)\n",
    "df_snv.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>INFO</th>\n",
       "      <th>sample_control</th>\n",
       "      <th>sample_tumor</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>genotype</th>\n",
       "      <th>TCN</th>\n",
       "      <th>PID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>842825</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>BRF=0.21;FR=1;HP=2;HapScore=1;MGOF=16;MMLQ=16;...</td>\n",
       "      <td>1/1:-103.9,-9.93,0:16:99:33:33</td>\n",
       "      <td>1/1:-69.7,-6.3,0:12:63:22:22</td>\n",
       "      <td>840009</td>\n",
       "      <td>12839977</td>\n",
       "      <td>1:1</td>\n",
       "      <td>2</td>\n",
       "      <td>4100314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>844300</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>BRF=0.8;FR=1;HP=3;HapScore=1;MGOF=46;MMLQ=32;M...</td>\n",
       "      <td>1/1:-55.7,-4.6,0:46:46:16:16</td>\n",
       "      <td>1/1:-50.5,-4.19,0:41:42:13:13</td>\n",
       "      <td>840009</td>\n",
       "      <td>12839977</td>\n",
       "      <td>1:1</td>\n",
       "      <td>2</td>\n",
       "      <td>4100314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>844324</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>BRF=0.65;FR=1;HP=2;HapScore=2;MGOF=60;MMLQ=15;...</td>\n",
       "      <td>1/1:-30.9,-3.27,0:60:33:14:14</td>\n",
       "      <td>1/1:-20.7,-2.37,0:52:24:11:10</td>\n",
       "      <td>840009</td>\n",
       "      <td>12839977</td>\n",
       "      <td>1:1</td>\n",
       "      <td>2</td>\n",
       "      <td>4100314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>844343</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>BRF=0.49;FR=1;HP=1;HapScore=4;MGOF=61;MMLQ=13;...</td>\n",
       "      <td>1/1:-27.5,-3.03,0:45:30:11:10</td>\n",
       "      <td>1/1:-17.5,-3.58,0:61:36:7:7</td>\n",
       "      <td>840009</td>\n",
       "      <td>12839977</td>\n",
       "      <td>1:1</td>\n",
       "      <td>2</td>\n",
       "      <td>4100314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>844476</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>BRF=0.81;FR=0.999;HP=4;HapScore=2;MGOF=13;MMLQ...</td>\n",
       "      <td>1/1:-3.4,-0,0:9:3:5:4</td>\n",
       "      <td>./.:-3.8,-0,0:13:3:1:1</td>\n",
       "      <td>840009</td>\n",
       "      <td>12839977</td>\n",
       "      <td>1:1</td>\n",
       "      <td>2</td>\n",
       "      <td>4100314</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   #CHROM     POS REF ALT                                               INFO   \n",
       "0       1  842825   A   G  BRF=0.21;FR=1;HP=2;HapScore=1;MGOF=16;MMLQ=16;...  \\\n",
       "1       1  844300   C   G  BRF=0.8;FR=1;HP=3;HapScore=1;MGOF=46;MMLQ=32;M...   \n",
       "2       1  844324   G   A  BRF=0.65;FR=1;HP=2;HapScore=2;MGOF=60;MMLQ=15;...   \n",
       "3       1  844343   A   G  BRF=0.49;FR=1;HP=1;HapScore=4;MGOF=61;MMLQ=13;...   \n",
       "4       1  844476   T   G  BRF=0.81;FR=0.999;HP=4;HapScore=2;MGOF=13;MMLQ...   \n",
       "\n",
       "                   sample_control                   sample_tumor   start   \n",
       "0  1/1:-103.9,-9.93,0:16:99:33:33   1/1:-69.7,-6.3,0:12:63:22:22  840009  \\\n",
       "1    1/1:-55.7,-4.6,0:46:46:16:16  1/1:-50.5,-4.19,0:41:42:13:13  840009   \n",
       "2   1/1:-30.9,-3.27,0:60:33:14:14  1/1:-20.7,-2.37,0:52:24:11:10  840009   \n",
       "3   1/1:-27.5,-3.03,0:45:30:11:10    1/1:-17.5,-3.58,0:61:36:7:7  840009   \n",
       "4           1/1:-3.4,-0,0:9:3:5:4         ./.:-3.8,-0,0:13:3:1:1  840009   \n",
       "\n",
       "        end genotype TCN      PID  \n",
       "0  12839977      1:1   2  4100314  \n",
       "1  12839977      1:1   2  4100314  \n",
       "2  12839977      1:1   2  4100314  \n",
       "3  12839977      1:1   2  4100314  \n",
       "4  12839977      1:1   2  4100314  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_snv.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_normal_genotype(df: pd.DataFrame) -> str:\n",
    "    gt = df[\"sample_control\"].split(\":\")[0]\n",
    "    if gt == \"1/0\" or gt == \"0/1\":\n",
    "        return \"1,1\"\n",
    "    if gt == \"0/0\":\n",
    "        return \"2,0\"\n",
    "    if gt == \"1/1\":\n",
    "        return \"0,2\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_snv[\"normal_genotype\"] = df_snv.apply(get_normal_genotype, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(3829904, 13)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_snv.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4100314\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0.82"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get purity for sample (PID)\n",
    "PID = df_snv.PID.values[0]\n",
    "print(PID)\n",
    "df_info = pd.read_csv(\"/home/junkhann/bioinf-d/Data/mmml/cnv_analysis/Identitymatrix_merged_additionalInfo.tsv\", delimiter=\"\\t\")\n",
    "purity = df_info.set_index(\"PID\").loc[str(PID), \"purity\"]\n",
    "purity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['1:1', nan, '2:1', '0:0', '1:0', 'sub:1', 'sub:sub'], dtype=object)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genotypes = df_snv[\"genotype\"].unique()\n",
    "genotypes\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_tumor_genotype_with_qualities(df: pd.DataFrame, purity: float) -> tuple[str, float]:\n",
    "    gt = df[\"genotype\"]\n",
    "    \n",
    "    if (gt != gt) or np.isnan(purity):\n",
    "        genotype = \"no_gt\"\n",
    "        quality_score = float('nan')\n",
    "    elif \"sub\" in gt:\n",
    "        genotype = \"sub\"\n",
    "        quality_score = float('nan')\n",
    "    else:\n",
    "        x = int(gt.split(\":\")[0])\n",
    "        y = int(gt.split(\":\")[1])\n",
    "        if x == y:\n",
    "            genotype = f\"{x},{y}\"\n",
    "            quality_score = 0\n",
    "        else:\n",
    "            try:\n",
    "                # maximum likelihood with n = NR, k = NV\n",
    "                n = int(df[\"sample_tumor\"].split(\":\")[-2])\n",
    "                k = int(df[\"sample_tumor\"].split(\":\")[-1])\n",
    "                tcn_tumor = float(df[\"TCN\"])\n",
    "\n",
    "                # case 1: x,y (ref, alt)\n",
    "                vaf_1 = (y * purity) / ((tcn_tumor * purity) + (2 * (1 - purity)))\n",
    "                prob_1 = binom.pmf(k=k, n=n, p=vaf_1)\n",
    "                \n",
    "                # case 2: y,x (ref, alt)\n",
    "                vaf_2 = (x * purity) / ((tcn_tumor * purity) + (2 * (1 - purity)))\n",
    "                prob_2 = binom.pmf(k=k, n=n, p=vaf_2)\n",
    "\n",
    "                # calculating raw phred-scaled likelihoods\n",
    "                pl_1 = -10 * math.log10(prob_1)\n",
    "                pl_2 = -10 * math.log10(prob_2)\n",
    "\n",
    "                # calculating quality score (capped at 99)\n",
    "                quality_score = min(99, abs(pl_1 - pl_2))\n",
    "\n",
    "                if prob_1 > prob_2:\n",
    "                    ref = x\n",
    "                    alt = y\n",
    "                else:\n",
    "                    ref = y\n",
    "                    alt = x\n",
    "                \n",
    "                genotype = f\"{ref},{alt}\"\n",
    "                \n",
    "            except ValueError:\n",
    "                genotype = \"no_gt\"\n",
    "                quality_score = float('nan')\n",
    "    \n",
    "    return (genotype, quality_score)\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(3829904, 15)"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_snv[\"tumor_genotype\"], df_snv[\"quality_score\"] = zip(*df_snv.apply(lambda x: get_tumor_genotype_with_qualities(x, purity), axis=1))\n",
    "df_snv.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_total_read_count_tumor(df: pd.DataFrame) -> int:\n",
    "    trc = df[\"sample_tumor\"].split(\":\")[-2]\n",
    "    return trc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_total_read_count_normal(df: pd.DataFrame) -> int:\n",
    "    trc = df[\"sample_control\"].split(\":\")[-2]\n",
    "    return trc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_snv[\"reads_normal\"] = df_snv.apply(get_total_read_count_normal, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_snv[\"reads_tumor\"] = df_snv.apply(get_total_read_count_tumor, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>INFO</th>\n",
       "      <th>sample_control</th>\n",
       "      <th>sample_tumor</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>genotype</th>\n",
       "      <th>TCN</th>\n",
       "      <th>PID</th>\n",
       "      <th>normal_genotype</th>\n",
       "      <th>tumor_genotype</th>\n",
       "      <th>quality_score</th>\n",
       "      <th>reads_normal</th>\n",
       "      <th>reads_tumor</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>156820</th>\n",
       "      <td>1</td>\n",
       "      <td>144920112</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>BRF=0.18;FR=0.5;HP=2;HapScore=1;MGOF=37;MMLQ=1...</td>\n",
       "      <td>1/0:-107.83,0,-82.43:37:99:71:39</td>\n",
       "      <td>1/0:-102.9,0,-170.9:30:99:93:35</td>\n",
       "      <td>144920087</td>\n",
       "      <td>145739956</td>\n",
       "      <td>2:1</td>\n",
       "      <td>3</td>\n",
       "      <td>4100314</td>\n",
       "      <td>1,1</td>\n",
       "      <td>2,1</td>\n",
       "      <td>27.545335</td>\n",
       "      <td>71</td>\n",
       "      <td>93</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156821</th>\n",
       "      <td>1</td>\n",
       "      <td>144920155</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>BRF=0.13;FR=0.5;HP=2;HapScore=1;MGOF=41;MMLQ=1...</td>\n",
       "      <td>0/1:-88.14,0,-74.54:41:99:60:34</td>\n",
       "      <td>0/1:-88.03,0,-145.73:33:99:83:32</td>\n",
       "      <td>144920087</td>\n",
       "      <td>145739956</td>\n",
       "      <td>2:1</td>\n",
       "      <td>3</td>\n",
       "      <td>4100314</td>\n",
       "      <td>1,1</td>\n",
       "      <td>2,1</td>\n",
       "      <td>20.535875</td>\n",
       "      <td>60</td>\n",
       "      <td>83</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156822</th>\n",
       "      <td>1</td>\n",
       "      <td>144920210</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>BRF=0.1;FR=0.5;HP=1;HapScore=1;MGOF=27;MMLQ=15...</td>\n",
       "      <td>1/0:-75.68,0,-70.18:27:99:58:29</td>\n",
       "      <td>1/0:-95.52,0,-146.22:23:99:87:37</td>\n",
       "      <td>144920087</td>\n",
       "      <td>145739956</td>\n",
       "      <td>2:1</td>\n",
       "      <td>3</td>\n",
       "      <td>4100314</td>\n",
       "      <td>1,1</td>\n",
       "      <td>2,1</td>\n",
       "      <td>3.192896</td>\n",
       "      <td>58</td>\n",
       "      <td>87</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156823</th>\n",
       "      <td>1</td>\n",
       "      <td>144920308</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>BRF=0.14;FR=0.5;HP=2;HapScore=2;MGOF=21;MMLQ=2...</td>\n",
       "      <td>1/0:-62.26,0,-36.36:21:99:53:24</td>\n",
       "      <td>1/0:-57.68,0,-93.58:16:99:72:25</td>\n",
       "      <td>144920087</td>\n",
       "      <td>145739956</td>\n",
       "      <td>2:1</td>\n",
       "      <td>3</td>\n",
       "      <td>4100314</td>\n",
       "      <td>1,1</td>\n",
       "      <td>2,1</td>\n",
       "      <td>32.442056</td>\n",
       "      <td>53</td>\n",
       "      <td>72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156824</th>\n",
       "      <td>1</td>\n",
       "      <td>144920425</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>BRF=0.21;FR=0.5;HP=4;HapScore=1;MGOF=36;MMLQ=1...</td>\n",
       "      <td>0/1:-59.24,0,-65.94:36:99:48:22</td>\n",
       "      <td>0/1:-52.07,0,-129.27:28:99:80:25</td>\n",
       "      <td>144920087</td>\n",
       "      <td>145739956</td>\n",
       "      <td>2:1</td>\n",
       "      <td>3</td>\n",
       "      <td>4100314</td>\n",
       "      <td>1,1</td>\n",
       "      <td>2,1</td>\n",
       "      <td>50.773895</td>\n",
       "      <td>48</td>\n",
       "      <td>80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3102442</th>\n",
       "      <td>14</td>\n",
       "      <td>106494503</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>BRF=0.13;FR=0.25;HP=3;HapScore=1;MGOF=10;MMLQ=...</td>\n",
       "      <td>1/0:-22.48,0,-47.28:5:99:24:9</td>\n",
       "      <td>0/0:0,-7.83,-84.9:10:78:28:0</td>\n",
       "      <td>106350732</td>\n",
       "      <td>106552286</td>\n",
       "      <td>1:0</td>\n",
       "      <td>1</td>\n",
       "      <td>4100314</td>\n",
       "      <td>1,1</td>\n",
       "      <td>1,0</td>\n",
       "      <td>99.000000</td>\n",
       "      <td>24</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3102449</th>\n",
       "      <td>14</td>\n",
       "      <td>106496138</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>BRF=0.19;FR=0.25;HP=5;HapScore=1;MGOF=13;MMLQ=...</td>\n",
       "      <td>1/0:-42.06,0,-69.26:13:99:38:13</td>\n",
       "      <td>0/0:0,-9.03,-104.7:10:90:33:0</td>\n",
       "      <td>106350732</td>\n",
       "      <td>106552286</td>\n",
       "      <td>1:0</td>\n",
       "      <td>1</td>\n",
       "      <td>4100314</td>\n",
       "      <td>1,1</td>\n",
       "      <td>1,0</td>\n",
       "      <td>99.000000</td>\n",
       "      <td>38</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3102466</th>\n",
       "      <td>14</td>\n",
       "      <td>106504370</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>BRF=0.18;FR=0.25;HP=4;HapScore=1;MGOF=22;MMLQ=...</td>\n",
       "      <td>0/1:-55.36,0,-62.66:22:99:42:21</td>\n",
       "      <td>0/0:0,-8.43,-98.5:10:84:28:0</td>\n",
       "      <td>106350732</td>\n",
       "      <td>106552286</td>\n",
       "      <td>1:0</td>\n",
       "      <td>1</td>\n",
       "      <td>4100314</td>\n",
       "      <td>1,1</td>\n",
       "      <td>1,0</td>\n",
       "      <td>99.000000</td>\n",
       "      <td>42</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3102467</th>\n",
       "      <td>14</td>\n",
       "      <td>106504413</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>BRF=0.23;FR=0.25;HP=1;HapScore=1;MGOF=16;MMLQ=...</td>\n",
       "      <td>1/0:-71.75,0,-71.25:16:99:50:27</td>\n",
       "      <td>0/0:0,-8.13,-92.4:8:81:27:0</td>\n",
       "      <td>106350732</td>\n",
       "      <td>106552286</td>\n",
       "      <td>1:0</td>\n",
       "      <td>1</td>\n",
       "      <td>4100314</td>\n",
       "      <td>1,1</td>\n",
       "      <td>1,0</td>\n",
       "      <td>99.000000</td>\n",
       "      <td>50</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3102489</th>\n",
       "      <td>14</td>\n",
       "      <td>106533084</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>BRF=0.16;FR=0.25;HP=2;HapScore=1;MGOF=11;MMLQ=...</td>\n",
       "      <td>0/1:-59.76,0,-54.56:9:99:40:20</td>\n",
       "      <td>0/0:0,-7.83,-89:11:78:24:0</td>\n",
       "      <td>106350732</td>\n",
       "      <td>106552286</td>\n",
       "      <td>1:0</td>\n",
       "      <td>1</td>\n",
       "      <td>4100314</td>\n",
       "      <td>1,1</td>\n",
       "      <td>1,0</td>\n",
       "      <td>99.000000</td>\n",
       "      <td>40</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>141297 rows Ã— 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         #CHROM        POS REF ALT   \n",
       "156820        1  144920112   G   A  \\\n",
       "156821        1  144920155   A   G   \n",
       "156822        1  144920210   G   A   \n",
       "156823        1  144920308   G   A   \n",
       "156824        1  144920425   C   T   \n",
       "...         ...        ...  ..  ..   \n",
       "3102442      14  106494503   G   A   \n",
       "3102449      14  106496138   G   A   \n",
       "3102466      14  106504370   C   T   \n",
       "3102467      14  106504413   C   A   \n",
       "3102489      14  106533084   C   T   \n",
       "\n",
       "                                                      INFO   \n",
       "156820   BRF=0.18;FR=0.5;HP=2;HapScore=1;MGOF=37;MMLQ=1...  \\\n",
       "156821   BRF=0.13;FR=0.5;HP=2;HapScore=1;MGOF=41;MMLQ=1...   \n",
       "156822   BRF=0.1;FR=0.5;HP=1;HapScore=1;MGOF=27;MMLQ=15...   \n",
       "156823   BRF=0.14;FR=0.5;HP=2;HapScore=2;MGOF=21;MMLQ=2...   \n",
       "156824   BRF=0.21;FR=0.5;HP=4;HapScore=1;MGOF=36;MMLQ=1...   \n",
       "...                                                    ...   \n",
       "3102442  BRF=0.13;FR=0.25;HP=3;HapScore=1;MGOF=10;MMLQ=...   \n",
       "3102449  BRF=0.19;FR=0.25;HP=5;HapScore=1;MGOF=13;MMLQ=...   \n",
       "3102466  BRF=0.18;FR=0.25;HP=4;HapScore=1;MGOF=22;MMLQ=...   \n",
       "3102467  BRF=0.23;FR=0.25;HP=1;HapScore=1;MGOF=16;MMLQ=...   \n",
       "3102489  BRF=0.16;FR=0.25;HP=2;HapScore=1;MGOF=11;MMLQ=...   \n",
       "\n",
       "                           sample_control                      sample_tumor   \n",
       "156820   1/0:-107.83,0,-82.43:37:99:71:39   1/0:-102.9,0,-170.9:30:99:93:35  \\\n",
       "156821    0/1:-88.14,0,-74.54:41:99:60:34  0/1:-88.03,0,-145.73:33:99:83:32   \n",
       "156822    1/0:-75.68,0,-70.18:27:99:58:29  1/0:-95.52,0,-146.22:23:99:87:37   \n",
       "156823    1/0:-62.26,0,-36.36:21:99:53:24   1/0:-57.68,0,-93.58:16:99:72:25   \n",
       "156824    0/1:-59.24,0,-65.94:36:99:48:22  0/1:-52.07,0,-129.27:28:99:80:25   \n",
       "...                                   ...                               ...   \n",
       "3102442     1/0:-22.48,0,-47.28:5:99:24:9      0/0:0,-7.83,-84.9:10:78:28:0   \n",
       "3102449   1/0:-42.06,0,-69.26:13:99:38:13     0/0:0,-9.03,-104.7:10:90:33:0   \n",
       "3102466   0/1:-55.36,0,-62.66:22:99:42:21      0/0:0,-8.43,-98.5:10:84:28:0   \n",
       "3102467   1/0:-71.75,0,-71.25:16:99:50:27       0/0:0,-8.13,-92.4:8:81:27:0   \n",
       "3102489    0/1:-59.76,0,-54.56:9:99:40:20        0/0:0,-7.83,-89:11:78:24:0   \n",
       "\n",
       "             start        end genotype TCN      PID normal_genotype   \n",
       "156820   144920087  145739956      2:1   3  4100314             1,1  \\\n",
       "156821   144920087  145739956      2:1   3  4100314             1,1   \n",
       "156822   144920087  145739956      2:1   3  4100314             1,1   \n",
       "156823   144920087  145739956      2:1   3  4100314             1,1   \n",
       "156824   144920087  145739956      2:1   3  4100314             1,1   \n",
       "...            ...        ...      ...  ..      ...             ...   \n",
       "3102442  106350732  106552286      1:0   1  4100314             1,1   \n",
       "3102449  106350732  106552286      1:0   1  4100314             1,1   \n",
       "3102466  106350732  106552286      1:0   1  4100314             1,1   \n",
       "3102467  106350732  106552286      1:0   1  4100314             1,1   \n",
       "3102489  106350732  106552286      1:0   1  4100314             1,1   \n",
       "\n",
       "        tumor_genotype  quality_score reads_normal reads_tumor  \n",
       "156820             2,1      27.545335           71          93  \n",
       "156821             2,1      20.535875           60          83  \n",
       "156822             2,1       3.192896           58          87  \n",
       "156823             2,1      32.442056           53          72  \n",
       "156824             2,1      50.773895           48          80  \n",
       "...                ...            ...          ...         ...  \n",
       "3102442            1,0      99.000000           24          28  \n",
       "3102449            1,0      99.000000           38          33  \n",
       "3102466            1,0      99.000000           42          28  \n",
       "3102467            1,0      99.000000           50          27  \n",
       "3102489            1,0      99.000000           40          24  \n",
       "\n",
       "[141297 rows x 17 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_snv[df_snv[\"quality_score\"] > 0 ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "93a303f0ef870f638483e4737d45cbeec6639f70307ce4ab90c35c8ca4d6e6b9"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
